src/lib/swapper/swappers/utils/helpers/helpers.ts

```diff
@@ -76,71 +58,42 @@ export const getERC20Allowance = ({
   return erc20Contract.methods.allowance(ownerAddress, spenderAddress).call()
 }
 
-export const isApprovalRequired = async ({
+export const getFeesFromContractData = async ({
+  accountNumber,
   adapter,
-  receiveAddress,
-  allowanceContract,
-  sellAsset,
-  sellAmountExcludeFeeCryptoBaseUnit,
-  web3,
-  erc20AllowanceAbi,
-}: IsApprovalRequiredArgs): Promise<boolean> => {
-  try {
-    if (sellAsset.assetId === adapter.getFeeAssetId()) {
-      return false
-    }
-
-    const ownerAddress = receiveAddress
-    const spenderAddress = allowanceContract
-
-    const { assetReference: sellAssetErc20Address } = fromAssetId(sellAsset.assetId)
-
-    const allowanceOnChain = await getERC20Allowance({
-      web3,
-      erc20AllowanceAbi,
-      ownerAddress,
-      spenderAddress,
-      sellAssetErc20Address,
-    })
-    if (!allowanceOnChain) {
-      throw new SwapError(`[isApprovalRequired] - No allowance data`, {
-        details: { allowanceContract, receiveAddress },
-        code: SwapErrorType.RESPONSE_ERROR,
-      })
-    }
-
-    if (bn(allowanceOnChain).isZero()) return true
-
-    const allowanceRequired = bnOrZero(sellAmountExcludeFeeCryptoBaseUnit).minus(allowanceOnChain)
-    return allowanceRequired.gt(0)
-  } catch (e) {
-    if (e instanceof SwapError) throw e
-    throw new SwapError('[isApprovalRequired]', {
-      cause: e,
-      code: SwapErrorType.ALLOWANCE_REQUIRED_FAILED,
-    })
-  }
-}
-
-export const getFeesFromFeeData = async ({
+  erc20ContractAddress,
+  erc20ContractData,
   wallet,
-  feeData,
-}: GetFeesFromFeeDataArgs): Promise<evm.Fees & { gasLimit: string }> => {
+}: GetFeesFromContractDataArgs): Promise<evm.Fees & { gasLimit: string }> => {
   if (!supportsETH(wallet)) {
-    throw new SwapError('[getFeesFromFeeData]', {
+    throw new SwapError('[getFeesFromContractData]', {
       cause: 'eth wallet required',
       code: SwapErrorType.SIGN_AND_BROADCAST_FAILED,
       details: { wallet },
     })
   }
 
-  const gasLimit = feeData.estimatedGasCryptoBaseUnit
-  const gasPrice = feeData.gasPriceCryptoBaseUnit
-  const maxFeePerGas = feeData.maxFeePerGas
-  const maxPriorityFeePerGas = feeData.maxPriorityFeePerGas
+  const ownerAddress = await adapter.getAddress({ accountNumber, wallet })
+
+  const { gasPrice, gasLimit, maxFeePerGas, maxPriorityFeePerGas } = await (async () => {
+    const getFeeDataInput = {
+      to: erc20ContractAddress,
+      value: '0',
+      chainSpecific: { from: ownerAddress, contractData: erc20ContractData },
+    }
+
+    // account for l1 transaction fees for optimism
+    if (optimism.isOptimismChainAdapter(adapter)) {
+      const { average, l1GasPrice } = await adapter.getFeeData(getFeeDataInput)
+      return Object.assign(average.chainSpecific, { gasPrice: l1GasPrice })
+    }
```

Comment:
appreciate feedback on the validity of this - this was originally found in thor swapper

Reply:
the `l1GasPrice` is handled by the node, so we shouldn't be using it for the `gasPrice` here. just pass through `average` fee data like any other evm chain. The only time we need to worry about l1 fees for optimism is for display purposes and balance check via `txFee` value

Reply:
roger that, ill fix

Reply:
oh damn @kaladinlight this object is used for tx AND display purposes so i'll move this concern into the UI layer

Reply:
Not sure I understand correctly but I've removed the custom logic for optimism gas fees. If this isnt right maybe we should catch up via pile

Reply:
> When functionally testing of approvals they worked as expected, and approval fee estimates are closely matching the actual.
> 
> <img alt="Screenshot 2023-05-17 at 3 18 52 pm" width="487" src="https://user-images.githubusercontent.com/97164662/238841681-08d9c8f3-7642-4b7b-b7d4-31b62b690041.png"> <img alt="Screenshot 2023-05-17 at 3 19 22 pm" width="474" src="https://user-images.githubusercontent.com/97164662/238841685-cefce139-117f-4373-abf1-23d7026c632b.png">
> However, I think we've broken something with our gas limits.
> 
> I attempted a THORSwap trade of ETH to AVAX, but the TX failed as the gas limit was too low. I attempted this twice to be sure, and both failed:
> 
> 1. https://etherscan.io/tx/0xa0824627d51a1ce0e16f8b21370b1352a6c16556a0ebb0e8f55ccc6312704817
> 2. https://etherscan.io/tx/0xd11067576be4ef029566cffd74a53282fa6a0c37194383d7b502ec5fcd9c67f6
> 
> When I attempted the same TX on production it was successful (note the significantly higher gas limit): https://etherscan.io/tx/0xcc05ec4707587b31402d7a8ed698bcd117539250bb7ca9b32832b2fd333786cd

well spotted ser. I've re-implemented hardcoded gas limit for thor EVM trades but also discovered how we can make it dynamic later, detailed in #4512 .

Reply:
https://etherscan.io/tx/0x82b4d7a2abfed1d7c76a10bc181e6c625c292700a4e152cc2da9a48390d405dc